src/odemis/driver/actuator.py

```diff
@@ -1366,3 +1366,510 @@ def terminate(self):
+        for child, caxis in zip(self._children, self._caxes_map):
+            f = child.reference({caxis})
+            futures.append(f)
+
+        # just wait for all futures to finish
+        for f in futures:
+            f.result()
+
+        # check if referencing pos matches any position allowed
+        pos_matching, pos_children = self._readPositionfromChildren()
+
+        # If we just did referencing and ended up to an unsupported position,
+        # move to any supported position
+        if not any(pos_matching):
+            # chose any position to move to
+            for key in self._positions:
+                self._doMoveAbs({self._axis_name: key})
+                break
+
+        # resubscribe after referencing
+        self._children_futures = []
+        for c in self._children:
+            c.position.subscribe(self._updatePosition)
+
+    def stop(self, axes=None):
+        """
+        stops the motion
+        axes (iterable or None): list of axes to stop, or None if all should be stopped
+        """
+        # Empty the queue for the given axes
+        if self._executor:
+            self._executor.cancel()
+
+        all_axes = set(self.axes.keys())
+        axes = axes or all_axes
+        unknown_axes = axes - all_axes
+        if unknown_axes:
+            logging.error("Attempting to stop unknown axes: %s", ", ".join(unknown_axes))
+            axes &= all_axes
+
+        threads = []
+        for child, ac in zip(self._children, self._caxes_map):
+            # it's synchronous, but we want to stop all of them as soon as possible
+            thread = threading.Thread(name="Stopping axis", target=child.stop, args=(ac,))
+            thread.start()
+            threads.append(thread)
+
+        # wait for completion
+        for thread in threads:
+            thread.join(1)
+            if thread.is_alive():
+                logging.warning("Stopping child actuator of '%s' is taking more than 1s", self.name)
+
+    def terminate(self):
+        if self._executor:
+            self.stop()
+            self._executor.shutdown()
+            self._executor = None
+
+        for c in self._children:
+            c.position.unsubscribe(self._updatePosition)
+        for c in [c for c, r in zip(self._children, self._children_refd) if r]:
+            c.referenced.unsubscribe(self._updateReferenced)
+
+
+class RotationActuator(model.Actuator):
+    """
+    A rotational actuator component which allows moving to any position
+    defined by the user upon initialization.
+    """
+
+    def __init__(self, name, role, children, axis_name, cycle=2*math.pi, inverted=None, **kwargs):
+        """
+        name (string)
+        role (string)
+        children (dict str -> actuator): axis name (in this actuator) -> actuator to be used for this axis
+        axis_name (str): axis name in the child actuator
+        cycle (float): this value represents a full cycle of the rotational actuator. If not specified a
```

Comment:
I'd say that cycle should be "0 < float", to keep things simple.

Reply:
Done